documentation/ArrayExprLoweringToFIR.md

```diff
@@ -0,0 +1,303 @@
+Note that it is unclear how well FIR transformation will be able to introduce short-circuits that are rather obvious looking at the `evaluate::Expr` tree, so this rule might get revisited.
+So far, the required short-circuits are:
+- `b.OR.SE`, `b.AND.SE` -> do not evaluate `SE` when `b` is false and `SE` has side effects.
+- `MERGE(SE1, SE2, b)` -> when `b` is a scalar, either `SE1` or `SE2` evaluation has side effects.
+Side effects should be understood in the broad sense (array accesses can cause segfaults, floating point operations can raise signals).
+##### Example
+In `b.AND.foo()` the call to `foo` shall not be performed if `b` is false.
+##### Plans to enforce the rule while lowering to FIR
+It is possible to use `fir.where` to implement a simple if/else logic without having to build a control flow graph using `br_cond` and basic blocks. Although both of these implementation are equivalent, `fir.where` gives more structural guarantees than basic block logic and is expected to interact in better way with optimizations. It is also more readable.
+However, with a `fir.where`, the result of such expression will need to be materialized (with basic blocks logic, one could have taken advantages of phi nodes scalar values, but FIR/LLVM transformation should anyway get rid of scalar temporary in such cases).
+The fir for the example above will look like:
+```
+%1 = load %b
+fir.where %1 {
+// b true case  
+  %1 = fir.call foo
+   fir.store %1 tmp 
+} otherwise {
+  %2 = constant 0
+  fir.store %2 tmp
+}
+```
+### Exposing enough structure so that array optimizations can be performed on FIR
+Because there are no FIR array optimizations yet, it is not entirely defined what this requirement implies yet. However, one of the objectives of MLIR is to be able to represent array layouts and access patterns as affine expressions. These affine expressions can then be composed allowing generalized loop and data transformations.
+Of course, this only make sense for array expressions that have such affine properties, mainly elemental and transformational operations. There is not much one can do to transform opaque function calls involving arrays into some affine operation.
+#### FIR operations to express array operations
+In FIR, affine expressions can be implemented using the `fir.loop`, `fir.where`, `fir.extract_value` and `fir.coordinate_of` operations in combinations with the `fir.array<>` type and the `fir.box<>` type that can hold affine maps encoding linear access patterns (see [FIR  Language reference](documentation/FIRLangRef.md)).
+#### Views over array in FIR
+In FIR, one can avoid copying array sections by creating `fir.box` with that holds the memory reference to the base variable as well as the subscript describing the re-mapping.
+For instance, `A(10:20:2, :)` where `A` is `REAL, DIMENSION(20, 10)` can be written in FIR as:
+```
+//%a is the memory reference to A storage, it is of type !fir.ref<fir.array<20:5:fir.real<4>>>
```

Comment:
Minor - Does FIR/MLIR use _x_ to separate triplet component values, rather than Fortran's _:_ separator?
I believe the extent of a 10:20:2 slice triplet is 6, rather than 5.
Some of the details for this example, such as references to _m_ and _n_, may be from a later example rather than this one.

Reply:
You are correct regarding all you points, thanks